core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -425,4 +447,28 @@ public boolean isClosed() {
     public Injector injector() {
         return this.injector;
     }
+
+    @SuppressForbidden(reason = "Need to specify port and pass to formatter")
+    private void writePortsFile(String type, BoundTransportAddress boundAddress) {
+        Path tmpPortsFile = environment.logsFile().resolve(type + ".ports.tmp");
+        try (BufferedWriter writer = Files.newBufferedWriter(tmpPortsFile, Charset.forName("UTF-8"))) {
+            for (TransportAddress address : boundAddress.boundAddresses()) {
+                InetSocketAddress socketAddress = new InetSocketAddress(address.getAddress(), address.getPort());
+                if (socketAddress.getAddress() instanceof Inet6Address &&
+                    socketAddress.getAddress().isLinkLocalAddress()) {
+                    // no link local, just causes problems
+                    continue;
+                }
+                writer.write(NetworkAddress.formatAddress(socketAddress) + "\n");
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to write ports file", e);
+        }
+        Path portsFile = environment.logsFile().resolve(type + ".ports");
+        try {
+            Files.move(tmpPortsFile, portsFile, StandardCopyOption.ATOMIC_MOVE);
```

Comment:
I've been wanting a portfile for a few weeks now for other testing things. It'll really make BWC tests more clean to have this, I think.